File: notes/0.0.3/Singleton.java
public class SingleObject {
    // 私有化构造函数，即外部不能使用 new SingleObject()，外部不能使用new!!
    private SingleObject(){}
    // 内部 new
    private SingleObject instance = null;
    // 对外接口
    public SingleObject getInstance() {
        if(instance === null) {
            // 保证只会 new 一次
            instance = new SingleObject();
        }
        return instance;
    }
    
    //对象方法
    public void show(name, pwd) {
        System.out.printIn('展示');
    }
}

public class SingleObjectDemo {
    public static void main(String[] args) {
        // 不合法
        SingleObject object = new SingleObject();

        // 正确使用，唯一可用可用对象
        SingleObject object = SingleObject.getInstance();
        object.show();
    }
}

File: notes/0.0.3/Singleton.js
class Singleton {
    constructor(name) {
        this.name = name;
        this.instance = null;
    }
    show() {
        console.log(this.name);
    }
}

Singleton.getInstance = function (name) {
    if (this.instance) {
        return this.instance;
    }
    return this.instance = new Singleton(name);
}

// 实例
// 只能使用静态函数 getInstance，不能使用 new Singleton()，但是只能文档约束
let s1 = Singleton.getInstance('展示1');
s1.show();
let s2 = Singleton.getInstance('展示2');
s2.show();
console.log(s1 === s2); // true


let s3 = new Singleton('展示3');
s3.show();
let s4 = new Singleton('展示4');
s4.show();
console.log(s3 === s4); // false